strawberryfields/compilers/gaussian_merge.py

```diff
@@ -0,0 +1,357 @@
+
+        Args:
+            seq (Sequence[Command]): quantum circuit to modify
+            registers (Sequence[RegRefs]): quantum registers
+        Returns:
+            List[Command]: modified circuit
+        """
+        self.curr_seq = seq
+        while self.merge_a_gaussian_op(registers):
+            continue
+        return self.curr_seq
+
+    def merge_a_gaussian_op(self, registers):
+        """
+        Main function to merge a gaussian operation with its gaussian neighbors.
+        If merge is achieved, the method updates self.curr_seq and returns True
+        Else (merge cannot be achieved), the method returns false
+        """
+        self.DAG = pu.list_to_DAG(self.curr_seq)
+
+        for op in list(self.DAG.nodes):
+            successors = list(self.DAG.successors(op))
+            predecessors = list(self.DAG.predecessors(op))
+            # If operation is a Gaussian operation
+            if get_op_name(op) in self.gaussian_ops:
+                merged_gaussian_ops = self.get_valid_gaussian_merge_ops(op)
+
+                # If there are successor operations that are Gaussian and can be merged
+                if merged_gaussian_ops:
+                    self.new_DAG = self.DAG.copy()
+                    # Fix order of operations
+                    merged_gaussian_ops = self.organize_merge_ops([op] + merged_gaussian_ops)
+                    gaussian_transform = GaussianUnitary().compile(merged_gaussian_ops, registers)
+                    self.new_DAG.add_node(gaussian_transform[0])
+
+                    # Logic to add displacement gates. Returns dictionary,
+                    # where the value is a displacement gate added and its key is the qumode its operating upon.
+                    displacement_mapping = self.add_displacement_gates(
+                        gaussian_transform, merged_gaussian_ops
+                    )
+
+                    # If there are predecessors: Attach predecessor edges to new gaussian transform
+                    if predecessors:
+                        self.new_DAG.add_edges_from(
+                            [(pre, gaussian_transform[0]) for pre in predecessors]
+                        )
+
+                    # Add edges to all successor operations not merged
+                    self.add_non_gaussian_successor_gates(
+                        gaussian_transform, successors, displacement_mapping
+                    )
+
+                    # Add edges for all successor/predecessor operations of the merged operations
+                    self.add_gaussian_pre_and_succ_gates(
+                        gaussian_transform, merged_gaussian_ops, displacement_mapping
+                    )
+
+                    self.new_DAG.remove_nodes_from([op] + merged_gaussian_ops)
+
+                    self.curr_seq = pu.DAG_to_list(self.new_DAG)
+                    return True
+        return False
+
+    def recursive_d_gate_successors(self, gate):
```

Comment:
Out of curiosity, how come the Dgate is merged separately compared to the GaussianTransform?

Reply:
The displacement gates never get actually merged into the gaussian transform, cause thats what Gaussian Unitary does. So I had to add logic to handle them. I put the after the gaussian transform and add edges to the successive operations that were not merged. Most of what made the code more complicated was trying to deal with displacement gates.

Reply:
Got it, thanks @federico0112